### ISSUE TITLE
Option to add and push git tags from the publish command

### ISSUE THREAD

**Issue by goodmorninggoaway** [lerna/lerna#2085]
I'd like an option to have `lerna publish from-package` to do the same git tagging as `lerna version`.

This may be related to #1951 .

 ## Expected Behavior
1. Add an option to the publish command to push tags, or
1. Add options to the version command to identify which packages will be published, but not bump the version

## Current Behavior
`lerna version` always changes the version unless you run in interactive mode and reset it to the same version. It tags the commit with the versions published at that commit. `lerna publish from-package` publishes but does not tag the commit.

## Possible Solution
Add a `--git-tag-version` option to the `publish from-package` command, which tag the commit. I may be able to submit a PR with this after next week.

## Context
 I'm trying to setup my project such that it publishes unpublished packages from CI on each commit to master. `lerna publish from-package` does everything I want _except_ the tagging. Next, I'm hoping to push release notes to GitHub which the `version` command does.

| Executable | Version |
| ---: | :--- |
| `lerna --version` | 3.13.1 |
| `npm --version`  | 6.9.0 |
| `yarn --version` | 1.13.0 |
| `node --version` | v8.15.0 |

| OS | Version |
| --- | --- |
| macOS Mojave | 10.14.4 |

**Comment by evocateur** [comment]
The only way `lerna publish from-package` even works is if you've used `lerna version` previously to bump package versions.

> Add an option to the publish command to push tags

That's what `lerna version` does (by default)

> Add options to the version command to identify which packages will be published

That's what `lerna changed` is for.

I'm very confused at what you're trying to do, here.

**Comment by tkissing-work** [comment]
I don't know what @goodmorninggoaway wants to do, but I want the same feature, so that I can have a Jenkins pipeline that does everything up to and including `lerna version` automatically (conventional commits, so no human input needed to figure out the version number), then waits for a human to approve.
After approval it should create/push the tag and publish to npm. No more, no less. 
And before you suggest it, no, I do not want to run `version` after the human approval because part of the approval process includes reviewing documentation that is generated during `version` and `postversion` hooks.

**Comment by goodmorninggoaway** [comment]
@evocateur I'm not using `lerna version` to bump package versions. I expect the version to be bumped by a human and reviewed in a PR (I don't want to start using conventional commits yet). When the PR is merged to `master`, CI will build and publish any changed packages. I have this working `lerna publish from-package --yes`, but the only thing I'm missing are tags.

Now that I write it out and based on what you've said, I could probably use `lerna changed` to determine the changes, then run `lerna version` to just reset the version and do everything else it does.

**Comment by goodmorninggoaway** [comment]
I thought I would be able to limit `lerna version` to specific pacakges, since I'm using independent versioning, but I can't.

**Comment by evocateur** [comment]
Nope, that's not what "independent" means. You run `lerna version`, it determines what packages changed since the last release (i.e., annotated git tag), and you get to choose which version to apply to each change ...independently. That's it.

**Comment by hardfist** [comment]
I also want the same feature, I don't know whether there's another way to achieve same goal。I want to do version&& publish in gitlab ci|cd, the following is my strategy
![image](https://user-images.githubusercontent.com/8898718/68266688-c7745c00-008a-11ea-98d6-f010748e4072.png)

1. version&& deploy is optional, when I click version, it will create to`lerna version --yes` which creates changelog and update version,but skipping push tag, and create a commit, then I create a mr using this commit (because we need to review this mr for fixing changelog text or fix version),so It's inappropriate to push tags in version steps.
2. after merge the previous mr, we can click deploy to publish to npm registry. which we want to push tags together

**Comment by iarroyo5** [comment]
Any update about this? I am also at the same situation as @tkissing-work . Have you found any work around about it?

**Comment by bigwoof91** [comment]
> I thought I would be able to limit `lerna version` to specific pacakges, since I'm using independent versioning, but I can't.

@goodmorninggoaway 

Although I'm extremely new to lerna, this crossed my mind as well while trying to setup publishing workflow for one of my (independent) packages! If `lerna publish from-package` exists then does it make sense to have `lerna version --scope`? Not "scope" necessarily, but some argument that is capable of running `version` for a  _**specific**_ package?

**Comment by goodmorninggoaway** [comment]
@bigwoof91 I ended up moving to a different project, so I’m not using let a anymore, but that seems promising if I ever get back to a lerna monorepo

**Comment by nickzelei** [comment]
This is something that would be really nice to have as well.

I have a monorepo that has N packages in it independently versioned.
Currently we have humans bump package versions (sometimes multiple in the same PR) and have CI actually do the publishing if it detects the versions have changed on a master commit.
It would be great to have lerna detect this via a flag and push annotated tags for those packages that will be published.

**Comment by jakewhelan** [comment]
We have the same process as @nickzelei. We manually bump versions and let CI publish for security reasons.

We're now building disaster recovery for our entire product, and that includes replaying publishing/rebuilding our npm registry from scratch by crawling commit history.

If CI could automatically tag after it publishes this would be far simpler for us, since we can just iterate over the tag list and re-publish.

**Comment by figo-hunter** [comment]
Well, I am not a frontend developer. I am a Unity3d developer. We are developing serveral upm packages within one Unity project. For prerequisite, upm is a Unity variant of npm; It has the same package.json definition; Packages are installed in /packages as well, And upm packages can be published to npm server. Unity built-in functionalities can resolve the name, version and dependency. The point is that we would like to have the package versioned by Unity developers whoes don't have node environment. For each project, lerna.json and package.json is autogenerated. And lerna publish is triggered by jenkins pipeline, for tag creating and npm publishing. In this case, lerna version is unnecessary and generate confusions to unity developers. We would like to have tag creation with publish from-package as well.

**Comment by figo-hunter** [comment]
> Nope, that's not what "independent" means. You run `lerna version`, it determines what packages changed since the last release (i.e., annotated git tag), and you get to choose which version to apply to each change ...independently. That's it.

Well, I am not a frontend developer. I am a Unity3d developer. We are developing serveral upm packages within one Unity project. For prerequisite, upm is a Unity variant of npm; It has the same package.json definition; Packages are installed in /packages as well, And upm packages can be published to npm server. Unity built-in functionalities can resolve the scopes, registries, names, versions and dependencies. The point is that we would like to have the package versioned by Unity developers whoes don't have node environment. For each project, lerna.json and package.json is autogenerated. And lerna publish is triggered by jenkins pipeline, for tag creating and npm publishing. In this case, lerna version is unnecessary and generate confusions to unity developers. We would like to have tag creation with publish from-package as well.

**Comment by yorjaggy** [comment]
I think most of the people here are asking for a feature to create the elegant tags lerna creates, but without publishing it
#2359 https://github.com/lerna/lerna/issues/2359

**Comment by william-will-angi** [comment]
We would also like this feature, however not for the purposes independently versioned packages. From an info-sec perspective, we can't allow anyone to have permissions to push to master (including a bot). Along the same vein, we can't allow any human to have permissions to push packages to the npm registry - We currently only expose credentials to a bot when running on the master branch.

Our use case is more similar to @tkissing-work and @inakiarroyo I think.

Our ideal workflow would be:
1. Developer makes commits using conventional commits and pushes to a merge/pull request
2. Those commits are merged in master.
3. Either through an automated CI/CD job or manually by another developer, `npm version --no-git-tag-version --no-push` is run to automatically generate changelogs and bump versions. Then those changes are added pushed to another MR/PR for human eyes to review.
4. The MR from step 3 merged into master. CI/CD pipelines uses `npm publish from-package` to publish any of those changed packages to our registry. 
5. (And this is the missing step) Based on those packages that were just published, create annotated git tags for each of them to signify a release.

Hoping maybe someone has found a decent workaround.
